File: examples/android/app/src/main/java/org/aetherfeed/app/sync/DriveAuth.kt
package org.aetherfeed.app.sync

import android.util.Base64
import java.security.MessageDigest
import java.security.SecureRandom

fun loopbackRedirect(): String = "http://127.0.0.1:$DRIVE_LOOPBACK_PORT"

fun createPkceVerifier(): String = b64url(randomBytes(32))

fun createPkceChallenge(verifier: String): String {
    val digest = MessageDigest.getInstance("SHA-256").digest(verifier.toByteArray(Charsets.US_ASCII))
    return b64url(digest)
}

fun buildAuthUrl(clientId: String, verifier: String): String {
    val challenge = createPkceChallenge(verifier)
    val params = listOf(
        "client_id" to clientId,
        "redirect_uri" to loopbackRedirect(),
        "response_type" to "code",
        "scope" to DRIVE_APPDATA_SCOPE,
        "code_challenge" to challenge,
        "code_challenge_method" to "S256",
        "access_type" to "offline",
        "prompt" to "consent",
    ).joinToString("&") { (k, v) ->
        "${java.net.URLEncoder.encode(k, "UTF-8")}=${java.net.URLEncoder.encode(v, "UTF-8")}"
    }
    return "https://accounts.google.com/o/oauth2/v2/auth?$params"
}

fun exchangeAuthCode(clientId: String, clientSecret: String, code: String, verifier: String): DriveTokens {
    val fields = mutableMapOf(
        "client_id" to clientId,
        "code" to code,
        "code_verifier" to verifier,
        "grant_type" to "authorization_code",
        "redirect_uri" to loopbackRedirect(),
    )
    if (clientSecret.isNotBlank()) fields["client_secret"] = clientSecret
    return parseTokens(formPost("https://oauth2.googleapis.com/token", fields))
}

fun refreshDriveToken(clientId: String, clientSecret: String, refreshToken: String): DriveTokens {
    val fields = mutableMapOf(
        "client_id" to clientId,
        "grant_type" to "refresh_token",
        "refresh_token" to refreshToken,
    )
    if (clientSecret.isNotBlank()) fields["client_secret"] = clientSecret
    val next = parseTokens(formPost("https://oauth2.googleapis.com/token", fields))
    return next.copy(refreshToken = next.refreshToken ?: refreshToken)
}

private fun parseTokens(obj: org.json.JSONObject): DriveTokens {
    val access = obj.optString("access_token")
    require(access.isNotBlank()) { "Drive token missing" }
    return DriveTokens(
        accessToken = access,
        refreshToken = obj.optString("refresh_token").ifBlank { null },
        expiresAt = System.currentTimeMillis() + obj.optLong("expires_in", 3600) * 1000,
    )
}

private fun randomBytes(size: Int): ByteArray = ByteArray(size).also { SecureRandom().nextBytes(it) }

private fun b64url(bytes: ByteArray): String =
    Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)

File: examples/web/src/sync/driveAuth.ts
import { DRIVE_APPDATA_SCOPE, DRIVE_LOOPBACK_PORT } from "@aetherfeed/domain";

export type DriveTokens = {
  accessToken: string;
  refreshToken?: string;
  expiresAt: number;
};

export function b64url(bytes: Uint8Array): string {
  let bin = "";
  for (const byte of bytes) bin += String.fromCharCode(byte);
  return btoa(bin).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/g, "");
}

export function createPkceVerifier(): string {
  return b64url(crypto.getRandomValues(new Uint8Array(32)));
}

export async function createPkceChallenge(verifier: string): Promise<string> {
  const digest = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(verifier));
  return b64url(new Uint8Array(digest));
}

export function loopbackRedirect(): string {
  return `http://127.0.0.1:${DRIVE_LOOPBACK_PORT}`;
}

export async function buildAuthUrl(clientId: string, verifier: string): Promise<string> {
  const challenge = await createPkceChallenge(verifier);
  const params = new URLSearchParams({
    client_id: clientId,
    redirect_uri: loopbackRedirect(),
    response_type: "code",
    scope: DRIVE_APPDATA_SCOPE,
    code_challenge: challenge,
    code_challenge_method: "S256",
    access_type: "offline",
    prompt: "consent",
  });
  return `https://accounts.google.com/o/oauth2/v2/auth?${params.toString()}`;
}

export function parseTokenResponse(data: Record<string, unknown>, now = Date.now()): DriveTokens {
  const access = String(data.access_token ?? "");
  if (!access) throw new Error("Drive token missing");
  return {
    accessToken: access,
    refreshToken: data.refresh_token ? String(data.refresh_token) : undefined,
    expiresAt: now + Number(data.expires_in ?? 3600) * 1000,
  };
}

export async function exchangeAuthCode(
  clientId: string,
  clientSecret: string,
  code: string,
  verifier: string,
): Promise<DriveTokens> {
  const body = new URLSearchParams({
    client_id: clientId,
    code,
    code_verifier: verifier,
    grant_type: "authorization_code",
    redirect_uri: loopbackRedirect(),
  });
  if (clientSecret) body.set("client_secret", clientSecret);
  const res = await fetch("https://oauth2.googleapis.com/token", {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body,
  });
  if (!res.ok) throw new Error(`token exchange failed (${res.status})`);
  return parseTokenResponse((await res.json()) as Record<string, unknown>);
}

export async function refreshDriveToken(
  clientId: string,
  clientSecret: string,
  refreshToken: string,
): Promise<DriveTokens> {
  const body = new URLSearchParams({
    client_id: clientId,
    grant_type: "refresh_token",
    refresh_token: refreshToken,
  });
  if (clientSecret) body.set("client_secret", clientSecret);
  const res = await fetch("https://oauth2.googleapis.com/token", {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body,
  });
  if (!res.ok) throw new Error(`token refresh failed (${res.status})`);
  const next = parseTokenResponse((await res.json()) as Record<string, unknown>);
  return { ...next, refreshToken: next.refreshToken ?? refreshToken };
}
